master/internal/resourcemanagers/agent/agent_state.go

```diff
@@ -0,0 +1,375 @@
+			continue
+		}
+
+		s.container = &msg.Container
+		if msg.Container.State == cproto.Terminated {
+			s.container = nil
+		}
+	}
+}
+
+func (a *AgentState) startContainer(ctx *actor.Context, msg sproto.StartTaskContainer) error {
+	inner := func(deviceId device.ID) error {
+		s, ok := a.slotStates[deviceId]
+		if !ok {
+			return errors.New("can't find slot")
+		}
+
+		// TODO(ilia): Potential race condition if slot is disabled in-between scheduling?
+		if !s.enabled.Enabled() {
+			return errors.New("container allocated but slot is not enabled")
+		}
+		if s.container != nil {
+			return errors.New("container already allocated to slot")
+		}
+
+		s.container = &msg.StartContainer.Container
+
+		return nil
+	}
+
+	for _, d := range msg.StartContainer.Container.Devices {
+		if err := inner(d.ID); err != nil {
+			return errors.Wrapf(err, "bad startedContainer on device: %d (%s)", d.ID, a.string())
+		}
+	}
+
+	a.containers[msg.Container.ID] = msg.TaskActor
+
+	return nil
+}
+
+func (a *AgentState) getSlotsSummary(ctx *actor.Context) model.SlotsSummary {
+	summary := make(model.SlotsSummary, len(a.slotStates))
+	for deviceID, slotState := range a.slotStates {
+		summary[fmt.Sprintf("%s/slots/%d", ctx.Self().Address(), deviceID)] = slotState.summarize()
+	}
+
+	return summary
+}
+
+func (a *AgentState) updateSlotDeviceView(ctx *actor.Context, deviceID device.ID) {
+	s, ok := a.slotStates[deviceID]
+	if !ok {
+		ctx.Log().Warnf("bad updateSlotDeviceView on device: %d (%s): not found", deviceID, a.string())
+		return
+	}
+
+	// TODO(ilia): Don't materialize `Devices` view on slots.
+	if s.enabled.Enabled() && !s.enabled.deviceAdded {
+		s.enabled.deviceAdded = true
+
+		var containerID *cproto.ID
+		if s.container != nil {
+			containerID = &s.container.ID
+		}
+
+		a.addDevice(ctx, s.device, containerID)
+	} else if !s.enabled.Enabled() {
+		if !s.enabled.draining && s.enabled.deviceAdded {
+			s.enabled.deviceAdded = false
+			a.removeDevice(ctx, s.device)
+		}
+
+		// On `PostStop`, draining will be already set to false, and we'll kill the container
+		// whether we have the device or not.
+		if !s.enabled.draining && s.container != nil {
+			ctx.Self().System().TellAt(s.container.Parent, task.Kill)
+		}
+	}
+}
+
+func (a *AgentState) patchSlotStateInner(
+	ctx *actor.Context, msg PatchSlotState, slotState *slot) model.SlotSummary {
+	if msg.Enabled != nil {
+		slotState.enabled.userEnabled = *msg.Enabled
+	}
+	if msg.Drain != nil {
+		slotState.enabled.draining = *msg.Drain
+	}
+	a.updateSlotDeviceView(ctx, slotState.device.ID)
+
+	return slotState.summarize()
+}
+
+func (a *AgentState) patchAllSlotsState(
+	ctx *actor.Context, msg PatchAllSlotsState) model.SlotsSummary {
+	result := model.SlotsSummary{}
+	for _, slotState := range a.slotStates {
+		summary := a.patchSlotStateInner(
+			ctx, PatchSlotState{ID: -1, Enabled: msg.Enabled, Drain: msg.Drain}, slotState)
```

Comment:
i get the -1 isn't used by it's still really confusing to read. why not just pass `slotState.device.id`